### PR TITLE
[No QA] Node-integration SQLite tests via better-sqlite3

### PR DESCRIPTION
**Pull request by sosek108**
### Details

Adds Node-side integration tests for `SQLiteProvider` so the SQLite storage path is covered in CI on every PR, instead of relying on manual device runs to catch SQL-shaped regressions. Until now, the provider had no real test coverage: unit tests stub out the database, and bugs in query strings, parameter binding, merge semantics, or escaping only surfaced once the code hit a real engine on device — late, expensive, and easy to miss. Running the actual provider against a real SQL engine in Node closes that gap at zero infrastructure cost: failures show up in seconds on the same CI that already runs the rest of the suite, refactors to the SQL layer become safe to make, and the contract between `SQLiteProvider` and the underlying database (merge behavior, null handling, injection safety, batch semantics) is now pinned down by executable tests rather than tribal knowledge.

### Related Issues

https://github.com/Expensify/App/issues/91861

### Linked E/App PR

https://github.com/Expensify/App/pull/92085

### Automated Tests

This PR introduces integration tests in `SQLiteProviderTest.ts`. This is new file and contains same tests as `IDBKeyvalProviderTest.ts` for different provider.

### Manual Tests

None. This PR adds automated testing only.

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I linked the corresponding Expensify/App PR in the `### Linked E/App PR` section above, and verified this change against it (E/App CI passed and manual testing completed)
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

### CODE REVIEW

File: tests/unit/mocks/sqliteMock.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+ *   - open({name})
+ *   - enableSimpleNullHandling()
+ *   - connection.execute(sql)
+ *   - connection.executeAsync<T>(sql, params?)
+ *   - connection.executeBatchAsync([{query, params}])
+ *
+ * Result rows are shaped to match Nitro: `{rows: {_array, item, length}}`.
+ */
+import BetterSqlite3 from 'better-sqlite3';
+import type {Database} from 'better-sqlite3';
```
[the remark this answers is not in the document]
Comment (pecanoro) [reply tests/unit/mocks/sqliteMock.ts:15]: @sosek108 Did you take a look at this comment?

[the remark this answers is not in the document]
Comment (Julesssss) [reply tests/unit/mocks/sqliteMock.ts:15]: It's not clear but I think the suggestion is `type Database = BetterSqlite3.Database;`. Or if this isn't a problem and works with `npm typecheck` please let us know.

File: tests/unit/storage/providers/SQLiteProviderTest.ts [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+            await SQLiteProvider.clear();
+            expect((await SQLiteProvider.getAllKeys()).length).toEqual(0);
+        });
+    });
+
+    describe('getDatabaseSize', () => {
+        it('should get the current size of the store', async () => {
+            await SQLiteProvider.setItem(ONYXKEYS.TEST_KEY, {payload: 'x'.repeat(1024)});
+            const size = await SQLiteProvider.getDatabaseSize();
+            expect(size.bytesUsed).toBeGreaterThan(0);
```
Comment (Julesssss) [inline tests/unit/storage/providers/SQLiteProviderTest.ts]: Apparently size will be non-0 after init anyway. Maybe we could compare the beofr and after size to verify it has increased?

File: tests/unit/storage/providers/SQLiteProviderTest.ts [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        // SQLite-specific regression: `multiSet` substitutes null for undefined
+        // before serializing, otherwise JSON.stringify(undefined) === undefined
+        // and the row would store a literal "undefined" string.
+        it('treats undefined as null', async () => {
+            await SQLiteProvider.multiSet([[ONYXKEYS.TEST_KEY, undefined as unknown as null]]);
+            expect(await SQLiteProvider.getItem(ONYXKEYS.TEST_KEY)).toBeNull();
+        });
+    });
+
+    describe('multiMerge', () => {
```
Comment (Julesssss) [inline tests/unit/storage/providers/SQLiteProviderTest.ts:135]: Could you add a multiMerge test where top level and nested keys are deleted when the value is null please

File: tests/unit/mocks/sqliteMock.ts [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    // prepare more than one statement at a time. SQLiteProvider's init() issues
+    // each statement separately, so this branch is rarely hit, but keep it
+    // defensive.
+    const semicolons = (sql.match(/;/g) ?? []).length;
+    if (semicolons > 1 || (semicolons === 1 && !sql.trim().endsWith(';'))) {
+        db.exec(sql);
+        return {rowsAffected: 0};
+    }
+
+    const {stmt, args} = prepareAndBind(db, sql, params);
```
Comment (pecanoro) [inline tests/unit/mocks/sqliteMock.ts]: I see we are using short names for some variables in this file. Could we use the long ones: `arguments`, `statement`, `connection`, or any others I missed?

**Review by pecanoro — CHANGES_REQUESTED** [review]

**Comment by fabioh8010** [comment]
Ssving to review tomorrow 👀

File: package-lock.json
Comment (fabioh8010) [inline package-lock.json:1]: Why do we have so many `"peer": true,` changes here? When running `npm install` locally I'm having changes here again, looks like "reverting" these `"peer": true,` changes

File: tests/unit/mocks/sqliteMock.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (fabioh8010) [reply tests/unit/mocks/sqliteMock.ts:15]: I tested locally too and I'm note seeing any issues with TS, CI is also passing. @sosek108 could you confirm?

File: tests/unit/mocks/sqliteMock.ts [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+type NitroResult<T> = {
+    rows?: NitroRows<T>;
+    rowsAffected: number;
+    insertId?: number;
+};
+
+type BatchQueryCommand = {
+    query: string;
+    params?: unknown[][];
+};
```
Comment (fabioh8010) [inline tests/unit/mocks/sqliteMock.ts]:
```suggestion
```

Let's use typings from `react-native-nitro-sqlite` instead to ensure correctness, see the next comments.

File: tests/unit/mocks/sqliteMock.ts [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        const name = m.slice(1);
+        if (!seen.has(name)) {
+            seen.add(name);
+            order.push(name);
+        }
+    }
+    return order;
+}
+
+function wrapRows<T extends Row>(rowsArray: T[]): NitroRows<T> {
```
Comment (fabioh8010) [inline tests/unit/mocks/sqliteMock.ts]:
```suggestion
function wrapRows<TRow extends QueryResultRow>(rowsArray: TRow[]): NitroSQLiteQueryResultRows<TRow> {
```

File: tests/unit/mocks/sqliteMock.ts [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    return {
+        _array: rowsArray,
+        item: (i: number) => rowsArray[i],
+        length: rowsArray.length,
+    };
+}
+
+function prepareAndBind(db: Database, sql: string, params: unknown[]) {
```
Comment (fabioh8010) [inline tests/unit/mocks/sqliteMock.ts]:
```suggestion
function prepareAndBind(db: Database, sql: string, params?: SQLiteQueryParams) {
```

File: tests/unit/mocks/sqliteMock.ts [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        const bindings: Record<string, unknown> = {};
+        for (let i = 0; i < namedOrder.length; i++) {
+            bindings[namedOrder[i]] = params[i];
+        }
+        return {stmt, args: [bindings] as const};
+    }
+    return {stmt: db.prepare(sql), args: params};
+}
+
+function runOne<T extends Row>(db: Database, sql: string, params: unknown[] = []): NitroResult<T> {
```
Comment (fabioh8010) [inline tests/unit/mocks/sqliteMock.ts]:
```suggestion
function runOne<TRow extends QueryResultRow>(db: Database, sql: string, params?: SQLiteQueryParams): QueryResult<TRow> {
```

File: tests/unit/mocks/sqliteMock.ts [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    if (stmt.reader) {
+        const rows = stmt.all(...(args as unknown[])) as T[];
+        return {rows: wrapRows(rows), rowsAffected: 0};
+    }
+
+    const info = stmt.run(...(args as unknown[]));
+    return {rowsAffected: info.changes, insertId: Number(info.lastInsertRowid)};
+}
+
+function makeConnection(name: string) {
```
Comment (fabioh8010) [inline tests/unit/mocks/sqliteMock.ts]:
```suggestion
function makeConnection(name: string): Pick<NitroSQLiteConnection, 'execute' | 'executeAsync' | 'executeBatchAsync' | 'close'> {
```

NAB but we should probably make mocks for the rest of the functions too, you can check the missing ones by doing this

```
function makeConnection(name: string): NitroSQLiteConnection {
```

File: tests/unit/mocks/sqliteMock.ts [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    if (!db) {
+        db = new BetterSqlite3(':memory:');
+        databases.set(name, db);
+    }
+    const conn = db;
+
+    return {
+        execute<T extends Row = Row>(sql: string, params: unknown[] = []): NitroResult<T> {
+            return runOne<T>(conn, sql, params);
+        },
```
Comment (fabioh8010) [inline tests/unit/mocks/sqliteMock.ts]:
```suggestion
        execute(sql, params) {
            return runOne(conn, sql, params);
        },
```

Simplified as we are typing `makeConnection` correctly now

File: tests/unit/mocks/sqliteMock.ts [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+
+        executeAsync<T extends Row = Row>(sql: string, params: unknown[] = []): Promise<NitroResult<T>> {
+            try {
+                return Promise.resolve(runOne<T>(conn, sql, params));
+            } catch (e) {
+                return Promise.reject(e);
+            }
+        },
```
Comment (fabioh8010) [inline tests/unit/mocks/sqliteMock.ts]:
```suggestion
        executeAsync(sql, params) {
            try {
                return Promise.resolve(runOne(conn, sql, params));
            } catch (e) {
                return Promise.reject(e);
            }
        },
```

Same

File: tests/unit/mocks/sqliteMock.ts [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+
+        executeBatchAsync(commands: BatchQueryCommand[]): Promise<{rowsAffected: number}> {
```
Comment (fabioh8010) [inline tests/unit/mocks/sqliteMock.ts]:
```suggestion
        executeBatchAsync(commands) {
```

File: tests/unit/mocks/sqliteMock.ts [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+    };
+}
+
+function open({name}: {name: string}) {
+    return makeConnection(name);
+}
+
+function enableSimpleNullHandling() {
+    // no-op; better-sqlite3 already handles null naturally.
+}
```
Comment (fabioh8010) [inline tests/unit/mocks/sqliteMock.ts]:
```suggestion
```

I don't think we need this anymore, we can also remove from jestSetup.js

File: tests/unit/mocks/sqliteMock.ts [hunk 14]
```diff
@@ -0,0 +1,197 @@
+/**
+ * Mock for `react-native-nitro-sqlite` backed by `better-sqlite3`, enabling
+ * Node-level integration tests against a real SQLite engine.
+ *
+ * Implements the NitroSQLite surface used by
+ * `lib/storage/providers/SQLiteProvider.ts`:
+ *   - open({name})
+ *   - enableSimpleNullHandling()
```
Comment (fabioh8010) [inline tests/unit/mocks/sqliteMock.ts]:
```suggestion
```

File: tests/unit/storage/providers/SQLiteProviderTest.ts [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+            await SQLiteProvider.multiMerge(changedEntries);
+
+            expect(await SQLiteProvider.getItem(ONYXKEYS.TEST_KEY)).toEqual('value_changed');
+            expect(await SQLiteProvider.getItem(ONYXKEYS.TEST_KEY_2)).toEqual(1001);
+            expect(await SQLiteProvider.getItem(ONYXKEYS.TEST_KEY_3)).toEqual(expectedTestKey3Value);
+            expect(await SQLiteProvider.getItem(`${ONYXKEYS.COLLECTION.TEST_KEY}id1`)).toEqual(false);
+            expect(await SQLiteProvider.getItem(`${ONYXKEYS.COLLECTION.TEST_KEY}id2`)).toEqual(['a', {newKey: 'newValue'}]);
+        });
+
+        it('inserts a new record when key does not exist', async () => {
```
Comment (fabioh8010) [inline tests/unit/storage/providers/SQLiteProviderTest.ts]:
```suggestion
        it('should insert a new record when key does not exist', async () => {
```

File: tests/unit/storage/providers/SQLiteProviderTest.ts [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+            await SQLiteProvider.multiMerge([[ONYXKEYS.TEST_KEY_2, {fresh: true}]]);
+            expect(await SQLiteProvider.getItem(ONYXKEYS.TEST_KEY_2)).toEqual({fresh: true});
+        });
+
+        it('shallow-merges existing record_key value', async () => {
```
Comment (fabioh8010) [inline tests/unit/storage/providers/SQLiteProviderTest.ts]:
```suggestion
        it('should shallow-merge existing record_key value', async () => {
```

File: tests/unit/storage/providers/SQLiteProviderTest.ts [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+            await SQLiteProvider.setItem(ONYXKEYS.TEST_KEY_3, {a: 1, b: 2});
+            await SQLiteProvider.multiMerge([[ONYXKEYS.TEST_KEY_3, {b: 99, c: 3}]]);
+            expect(await SQLiteProvider.getItem(ONYXKEYS.TEST_KEY_3)).toEqual({a: 1, b: 99, c: 3});
+        });
+
+        it('deep-merges nested objects', async () => {
```
Comment (fabioh8010) [inline tests/unit/storage/providers/SQLiteProviderTest.ts]:
```suggestion
        it('should deep-merge nested objects', async () => {
```

File: tests/unit/storage/providers/SQLiteProviderTest.ts [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+            });
+            await SQLiteProvider.multiMerge([[ONYXKEYS.TEST_KEY_3, {outer: {b: 99, nested: {y: 99, z: 3}}}]]);
+            expect(await SQLiteProvider.getItem(ONYXKEYS.TEST_KEY_3)).toEqual({
+                outer: {a: 1, b: 99, nested: {x: 1, y: 99, z: 3}},
+            });
+        });
+
+        // SQLite-specific: the JSON_REPLACE path is what makes `REPLACE_OBJECT_MARK`
+        // actually wipe a nested object (JSON_PATCH alone would only merge into it).
+        it('fully replaces a nested object marked with REPLACE_OBJECT_MARK via JSON_REPLACE', async () => {
```
Comment (fabioh8010) [inline tests/unit/storage/providers/SQLiteProviderTest.ts]:
```suggestion
        it('should fully replace a nested object marked with REPLACE_OBJECT_MARK via JSON_REPLACE', async () => {
```

File: tests/unit/storage/providers/SQLiteProviderTest.ts [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+            await SQLiteProvider.removeItems([ONYXKEYS.TEST_KEY, ONYXKEYS.TEST_KEY_3]);
+            expect(await SQLiteProvider.getAllKeys()).not.toContain(ONYXKEYS.TEST_KEY);
+            expect(await SQLiteProvider.getAllKeys()).not.toContain(ONYXKEYS.TEST_KEY_3);
+        });
+    });
+
+    // SQLite-specific: the IN-list is parameterised, so a key containing SQL
+    // fragments must be treated as a literal record_key.
+    describe('SQL-injection safety', () => {
+        it('treats a key containing SQL fragments as a literal record_key', async () => {
```
Comment (fabioh8010) [inline tests/unit/storage/providers/SQLiteProviderTest.ts]:
```suggestion
        it('should treat a key containing SQL fragments as a literal record_key', async () => {
```

**Review by fabioh8010 — CHANGES_REQUESTED** [review]

File: tests/unit/mocks/sqliteMock.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (sosek108) [reply tests/unit/mocks/sqliteMock.ts:15]: for me current state works well. No issues from typecheck. I'd leave it as is

```
import type {Database} from 'better-sqlite3';
```

File: tests/unit/storage/providers/SQLiteProviderTest.ts (hunk 3 above)
Comment (sosek108) [reply tests/unit/storage/providers/SQLiteProviderTest.ts:135]: added **deletes top-level and nested keys when the merge value is null**

File: package-lock.json
Comment (sosek108) [reply package-lock.json:1]: Merged `main` and `peer` changes disappeared

File: tests/unit/storage/providers/SQLiteProviderTest.ts (hunk 2 above)
Comment (sosek108) [reply tests/unit/storage/providers/SQLiteProviderTest.ts]: modified `should report a larger bytesUsed after a write` test so it checks wheter `after` size is bigger than `before` size

File: tests/unit/mocks/sqliteMock.ts (hunk 4 above)
Comment (sosek108) [reply tests/unit/mocks/sqliteMock.ts]: Corrected names.

File: tests/unit/mocks/sqliteMock.ts (hunk 5 above)
Comment (sosek108) [reply tests/unit/mocks/sqliteMock.ts]: Applied 👍

File: tests/unit/mocks/sqliteMock.ts (hunk 9 above)
Comment (sosek108) [reply tests/unit/mocks/sqliteMock.ts]: why do you think we need other methods?

**Comment by sosek108** [comment]
@pecanoro @fabioh8010 @Julesssss thank you for the review. I think that I've applied all of your suggestions (except for one NAB from Fabio)

File: tests/unit/storage/providers/SQLiteProviderTest.ts [hunk 20]
```diff
@@ -0,0 +1,347 @@
...
+            });
+            await SQLiteProvider.multiMerge([[ONYXKEYS.TEST_KEY_3, {outer: {b: 99, nested: {y: 99, z: 3}}}]]);
+            expect(await SQLiteProvider.getItem(ONYXKEYS.TEST_KEY_3)).toEqual({
+                outer: {a: 1, b: 99, nested: {x: 1, y: 99, z: 3}},
+            });
+        });
+
+        // RFC 7396 (JSON Merge Patch): a `null` value in the patch removes the key
+        // from the target. SQLite's `JSON_PATCH` implements this directly.
+        it('deletes top-level and nested keys when the merge value is null', async () => {
```
Comment (fabioh8010) [inline tests/unit/storage/providers/SQLiteProviderTest.ts:197]:
```suggestion
        it('should delete top-level and nested keys when the merge value is null', async () => {
```

**Review by fabioh8010 — APPROVED** [review]
Minor nit to address, other than that LGTM 🚀

**Review by pecanoro — APPROVED** [review]

**Review by Julesssss — COMMENTED** [review]
Thanks for updates

**Review by Julesssss — APPROVED** [review]
